proof/refine/ARM/Tcb_R.thy

```diff
@@ -1963,768 +1943,1232 @@ where
-                          frame;
-                    cur \<leftarrow> gets cur_thread;
-                    liftE $ when (cur = a) (reschedule_required)
-                 od
-             odE))
-       g)
-     (case_option (returnOk ())
-        (\<lambda>(ptr, frame).
-            do bufferSlot \<leftarrow> getThreadBufferSlot a;
-            doE y \<leftarrow> cteDelete bufferSlot True;
-            do y \<leftarrow> threadSet (tcbIPCBuffer_update (\<lambda>_. ptr)) a;
-               y \<leftarrow> (case_option (return ())
-                      (case_prod
-                        (\<lambda>newCap srcSlot.
-                            checkCapAt newCap srcSlot $
-                            checkCapAt
-                             (capability.ThreadCap a)
-                             sl' $
-                            assertDerived srcSlot newCap $
-                            cteInsert newCap srcSlot bufferSlot))
-                      frame);
-               cur \<leftarrow> getCurThread;
-               liftE $ when (cur = a) rescheduleRequired
-            od odE od)
-        g')" (is "corres _ ?T2_pre ?T2_pre' _ _")
-    using z sl
-    apply -
-    apply (rule corres_guard_imp[where P=P and P'=P'
-                                  and Q="P and cte_at (a, tcb_cnode_index 4)"
-                                  and Q'="P' and cte_at' (cte_map (a, cap))" for P P' a cap])
-      apply (cases g)
-       apply (simp, simp add: returnOk_def)
-      apply (clarsimp simp: liftME_def[symmetric] U2 liftE_bindE)
-      apply (case_tac b, simp_all add: newroot_rel_def)
-       apply (rule corres_guard_imp)
-         apply (rule corres_split_norE)
-            apply (rule_tac F="is_aligned aa msg_align_bits" in corres_gen_asm2)
-            apply (rule corres_split_nor)
-               apply (rule corres_split [OF _ gct_corres], clarsimp)
-                 apply (rule corres_when[OF refl rescheduleRequired_corres])
-                apply (wpsimp wp: gct_wp)+
-              apply (rule threadset_corres,
-                      (simp add: tcb_relation_def), (simp add: exst_same_def)+)[1]
-             apply (wp hoare_drop_imp)
-             apply (rule threadcontrol_corres_helper1[unfolded pred_conj_def])
-            apply simp
-            apply (wp hoare_drop_imp)
-            apply (wp threadcontrol_corres_helper2 | wpc | simp)+
-           apply (rule cap_delete_corres)
-          apply wp
-         apply (wpsimp wp: cteDelete_invs' hoare_vcg_conj_lift)
-        apply (fastforce simp: emptyable_def)
-       apply fastforce
+    apply (subgoal_tac "obj_at' (\<lambda>sc. \<exists>y. scTCB sc = Some y) scp s'")
+     apply (clarsimp simp: sc_at_pred_def obj_at_def obj_at'_def)
+    apply (prop_tac "sc_at scp s")
+     apply (clarsimp simp: sc_at_pred_def obj_at_def obj_at'_def is_sc_obj_def)
+     apply (fastforce dest: invs_valid_objs simp: valid_obj_def)
+    apply (drule sc_at_cross[rotated 3], fastforce, fastforce, fastforce)
```

Comment:
It's not really any different, but this can be `apply (drule sc_at_cross[rotated 3]; fastforce?)` I think.

Reply:
Doesn't work here

Reply:
Condensed this line and the following lines into a single fastforce instance